pykafka/producer.py

```diff
@@ -317,30 +350,15 @@ def _get_partition_msgs(partition_id, req):
                 time.sleep(self._retry_backoff_ms / 1000)
                 self._prepare_request(to_retry, owned_broker, attempt)
             else:
-                raise ProduceFailureError('Unable to produce messages. See log for details.')
-
-    def _update_leaders(self):
-        """Ensure each message in each queue is in the queue owned by its
-            partition's leader
-
-        This function empties all broker queues, maps their messages to the
-        current leaders for their partitions, and enqueues the messages in
-        the appropriate queues.
-        """
-        # empty queues and figure out updated partition leaders
-        new_queue_contents = defaultdict(list)
-        for owned_broker in itervalues(self._owned_brokers):
-            owned_broker.lock.acquire()
-            current_queue_contents = owned_broker.flush(0, release_pending=True)
-            for kv, partition_id in current_queue_contents:
-                partition_leader = self._topic.partitions[partition_id].leader
-                new_queue_contents[partition_leader.id].append((kv, partition_id))
-        # retain locks for all brokers between these two steps
-        for owned_broker in itervalues(self._owned_brokers):
-            owned_broker.enqueue(new_queue_contents[owned_broker.broker.id],
-                                 self._block_on_queue_full)
-            owned_broker.resolve_event_state()
-            owned_broker.lock.release()
+                log.error("Failed to produce messages to broker %s:%s after %s attempts. "
+                          "Re-enqueuing %s messages.", owned_broker.broker.host,
+                          owned_broker.broker.port, attempt, len(to_retry))
+                owned_broker.increment_messages_pending(-1 * len(to_retry))
+                for kv, partition_id, msg_attempt in to_retry:
+                    if msg_attempt >= self._max_message_retries:
+                        raise ProduceFailureError("Message failed to send after %d "
+                                                  "retries.", self._max_message_retries)
+                    self._produce((kv, partition_id, msg_attempt + 1))
```

Comment:
Thanks for this update, that looks perfect to re-enable the `max_retries=0` option. And seeing how this works with the attempts-tracking, it suddenly "clicked" in my head, what you said about why L347-351 don't help much if a broker has gone down or a leader has changed.

I kept misreading those lines before, thinking the fact that there's now an `_update` in L339 should fix everything, but now I get that it would still resend to the old destination, as per `owned_broker` in L351.

So given that the former retry loop doesn't really work so well, and that this does, would it make sense to drop that other one altogether? So what I mean is, you'd be left with:
```python
         if to_retry:
                owned_broker.increment_messages_pending(-1 * len(to_retry))
                for kv, partition_id, msg_attempt in to_retry:
                    if msg_attempt >= self._max_retries:
                        raise ProduceFailureError("Message failed to send after %d "
                                                  "retries.", self._max_retries)
                    self._produce((kv, partition_id, msg_attempt + 1))
```
(with possibly some copy-paste errors, and I've replace `_max_message_retries` by `_max_retries`, so you'd be back to one parameter for this).